src/test/common/test_hobject.cc

```diff
@@ -9,3 +16,303 @@ TEST(HObject, cmp)
+	 "n5"s},
+     "0000000000002727.0F0F0E0E.12345678.nonprint\x18%p%u%p%e%p.c.n5",
+     "10023:0f0f0707:n5:c:nonprint%18%25_%25.%25:12345678"},
+
+    {hobject_t{object_t("o//////"), string("ZZ"), 0xaaaa, 65, 1, "zzzzz"},
+     "0000000000000001.14000000.aaaa.o//////.ZZ.zzzzz",
+     "1:82000000:zzzzz:ZZ:o%2f%2f%2f%2f%2f%2f:aaaa"}};
+
+// original Ceph code as it was in version Squid
+
+struct test_hobject_fmt_t : public hobject_t {
+
+  template <typename... ARGS>
+  test_hobject_fmt_t(ARGS&&... args) : hobject_t{std::forward<ARGS>(args)...}
+  {}
+
+  test_hobject_fmt_t(const test_hobject_fmt_t& rhs) = default;
+  test_hobject_fmt_t(test_hobject_fmt_t&& rhs) = default;
+  test_hobject_fmt_t& operator=(const test_hobject_fmt_t& rhs) = default;
+  test_hobject_fmt_t& operator=(test_hobject_fmt_t&& rhs) = default;
+  test_hobject_fmt_t(hobject_t_max&& singleton) : test_hobject_fmt_t()
+  {
+    max = true;
+  }
+  test_hobject_fmt_t& operator=(hobject_t_max&& singleton)
+  {
+    *this = hobject_t();
+    max = true;
+    return *this;
+  }
+  bool is_max() const { return max; }
+  bool is_min() const
+  {
+    // this needs to match how it's constructed
+    return snap == 0 && hash == 0 && !max && pool == INT64_MIN;
+  }
+
+  constexpr auto operator<=>(const test_hobject_fmt_t& rhs) const noexcept
+  {
+    auto cmp = is_max() <=> rhs.is_max();
+    if (cmp != 0)
+      return cmp;
+    cmp = pool <=> rhs.pool;
+    if (cmp != 0)
+      return cmp;
+    cmp = get_bitwise_key() <=> rhs.get_bitwise_key();
+    if (cmp != 0)
+      return cmp;
+    cmp = nspace <=> rhs.nspace;
+    if (cmp != 0)
+      return cmp;
+    if (!(get_key().empty() && rhs.get_key().empty())) {
+      cmp = get_effective_key() <=> rhs.get_effective_key();
+      if (cmp != 0)
+	return cmp;
+    }
+    cmp = oid <=> rhs.oid;
+    if (cmp != 0)
+      return cmp;
+    return snap <=> rhs.snap;
+  }
+  constexpr bool operator==(const hobject_t& rhs) const noexcept
+  {
+    return operator<=>(rhs) == 0;
+  }
+};
+
+static inline void append_out_escaped(const std::string& in, std::string* out)
+{
+  for (auto i = in.cbegin(); i != in.cend(); ++i) {
+    if (*i == '%' || *i == ':' || *i == '/' || *i < 32 || *i >= 127) {
+      char buf[4];
+      snprintf(buf, sizeof(buf), "%%%02x", (int)(unsigned char)*i);
+      out->append(buf);
+    } else {
+      out->push_back(*i);
+    }
+  }
+}
+
+// why don't we escape non-printable characters?
+static void append_escaped(const string& in, string* out)
+{
+  for (string::const_iterator i = in.begin(); i != in.end(); ++i) {
+    if (*i == '%') {
+      out->push_back('%');
+      out->push_back('p');
+    } else if (*i == '.') {
+      out->push_back('%');
+      out->push_back('e');
+    } else if (*i == '_') {
+      out->push_back('%');
+      out->push_back('u');
+    } else {
+      out->push_back(*i);
+    }
+  }
+}
+
+// original Ceph code as it was in version Squid
+string hobject_t::to_str() const
```

Comment:
Why these functions needs to be copied here and can't be re-used from `common/hobject.cc`?
Same for append_escaped, append_out_escaped

Reply:
Ah, this is the previous version of this code.

Reply:
Interesting addition to the tests, but not a replacement. But:
- I did not find such a ctor. Maybe I missed it? and
- there are (at least) two different ways to 'stringify' an hobject: to_str() and the formatter use
 different escaping mechanism (which is interesting by itself). 

I do want to keep a fixed base line, so there is a point in copying the code in.